core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/ShardsAllocators.java

```diff
@@ -76,7 +76,20 @@ public boolean allocateUnassigned(RoutingAllocation allocation) {
 
     @Override
     public boolean rebalance(RoutingAllocation allocation) {
-        return allocator.rebalance(allocation);
+        final int numberOfInFlightFetch = gatewayAllocator.getNumberOfInFlightFetch();
```

Comment:
I *think* we have a race condition here where the gateway's requests have come back before we got here on the cluster state update thread. The in flight ops will be 0 but the results are not used yet. I wonder if we should have a flag on the allocation, something like "disableRebalance" which we can set in the gateway allocator. Maybe that will be simpler?